### PR TITLE
Improve typings

### PR DESCRIPTION
**Pull request by unional**
I have added a few things:

- Add type guard
- Add generic types to the interface. The main usage is on `isFSA` and `isError`
- Add test

This allows users to specify the type of the `payload` and `meta` easily.
See `typings-test.ts` for examples.

In the future, I would like to make the `Meta` type optional.
But need to wait for https://github.com/Microsoft/TypeScript/issues/2175

For now, I think this is the best it can get. 🌷

### CODE REVIEW

File: test/typings-test.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import { FluxStandardAction, isError, isFSA } from '../src'
```
Comment (unional) [inline test/typings-test.ts]: Ok.

File: package.json [hunk 2]
```diff
@@ -14,7 +14,7 @@
     "lint": "eslint src/ test/",
     "prepublish": "npm test && npm run build",
     "pretest": "npm run lint",
-    "test": "cross-env NODE_ENV=test nyc mocha"
+    "test": "cross-env NODE_ENV=test nyc mocha && tsc"
```
Comment (unional) [inline package.json]: Nice, never use posttest before. :)

**Comment by unional** [comment]
Updated based on feedback. 🌷

**Comment by unional** [comment]
The last change I made worth debating. On one side it helps on the type guard usage, on the other side it makes action creation harder:
https://github.com/Microsoft/TypeScript/issues/12400

What do you think?

**Comment by unional** [comment]
I think it is the right change. When `payload` and `meta` is defined by a custom action, the action author would expect the payload and meta to be non-optional.

Would that be any case that this is not true?

**Comment by unional** [comment]
cc/ @tkqubo who add typings to DT.

**Comment by unional** [comment]
@JaKXz do you have anything for me to change before we can merge this? 🌷

**Comment by unional** [comment]
@JaKXz good to hear from you!

> here are some more nitpicks I could make in typings-test.ts about code style [e.g. missing semicolons 😱 ] - can that file [and other .ts files in the repo] be linted somehow?

It still have? Seems like I'm getting very comfortable with my own linting standard 😛 .
I can use `tslint` (which I do for my projects), but the key is getting the same linting config as your are comfortable with.
Let me know what do you want to fix (other than the semicolons), and I will fix them.


> the posttest action that runs tsc will return an exit code of 1 if there is a problem with the typings, correct?

Yes, but since it is `posttest`, I don't think it will stop the build.

> please run yarn to update yarn.lock

I personally do not recommend checking in `yarn.lock` for module library. It makes breakage detection harder for the maintainer. I can do it if you still want to.

> what is the status on the typescript issue

They are working on it, likely will still take a few months to land. We can update the typing when that is released and get mainstreamed.

**Comment by unional** [comment]
Found the comment I made about `yarn.lock`: https://github.com/greenkeeperio/greenkeeper/issues/314#issuecomment-270993958

**Comment by JaKXz** [comment]
perhaps https://github.com/nzakas/eslint-plugin-typescript would help?

**Comment by unional** [comment]
> it would be ideal if the tsc command exited with a non-zero code

It does exist with non-zero. I was just concerning that `posttest` non-zero exit will be ignored. Seems like it will not be the case.

> I think the purpose of the yarn.lock file is to give some reassurance about what versions of dependencies and sub-dependencies are resolved & installed

yes, that's true for applications. For modules/packages, IMO it will hide the problem until a user reports it. E.g. if your module depends on `moduleA@1.2.3` and unfortunately `moduleA@1.2.4` breaks your code, you won't know it until some users use your module and discover it is broken.

This is because during consumption, the `yarn.lock` file of your module is not used to lock sub-dependency.

The article (https://yarnpkg.com/blog/2016/11/24/lockfiles-for-all) about devDeps vs deps is correct thou.

All in all, maybe this topic is still in flux and it will clear out as more module author adopt one way or the other. This is just a friendly discussion. 🌷 

Maybe using `yarn.lock` with `greenkeeper` would work out just fine?

I have updated the `yarn.lock` file anyway, as you requested. 🌷

File: package.json [hunk 3]
```diff
@@ -43,12 +44,15 @@
...
+    "rimraf": "^2.5.4",
+    "typescript": "^2.0.10"
   },
   "dependencies": {
+    "eslint-plugin-typescript": "^0.1.0",
     "lodash.isplainobject": "^4.0.6",
     "lodash.isstring": "^4.0.1",
-    "lodash.issymbol": "^4.0.1"
+    "lodash.issymbol": "^4.0.1",
+    "typescript-eslint-parser": "^1.0.2"
```
Comment (JaKXz) [inline package.json]: I think all typescript dependencies should be in `devDeps`.

**Comment by JaKXz** [comment]
@unional thank you for your thoughts, hard work, and patience on this. I'm excited to get it merged after one _last_ issue and release v1.1.0 :)

File: package.json (hunk 3 above)
Comment (unional) [inline package.json]: Yes, of course. 😛

File: package.json [hunk 4]
```diff
@@ -43,9 +44,12 @@
     "eslint-plugin-import": "^2.2.0",
     "mocha": "^3.1.2",
     "nyc": "^10.0.0",
-    "rimraf": "^2.5.4"
+    "rimraf": "^2.5.4",
+    "typescript": "^2.0.10",
+    "typescript-eslint-parser": "^1.0.2"
   },
   "dependencies": {
+    "eslint-plugin-typescript": "^0.1.0",
```
Comment (JaKXz) [inline package.json]: missed this one as well

**Pull request merged**